### PR TITLE
[AFQMC] Implement uninit_copy for shm_ptr_with_raw_ptr_dispatch

### PR DESCRIPTION
**Pull request by correaa**
declare stub for uninitialized_copy_n of shm_ptr in AFQMC

## Proposed changes

Implement missing algorithm reflecting changes in Multi

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- Refactoring (no functional changes, no api changes)

### Does this introduce a breaking change?

- No

## Checklist

- Yes. This PR is up to date with current the current state of 'develop'
- No. Code added or changed in the PR has been clang-formatted
- No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

File: src/AFQMC/Memory/SharedMemory/shm_ptr_with_raw_ptr_dispatch.hpp [hunk 1]
```diff
@@ -480,6 +480,22 @@ shm_ptr_with_raw_ptr_dispatch<T> uninitialized_copy_n(It1 f, Size n, shm_ptr_wit
   return d + n;
 }
 
+template<class T, class Size, typename It2>
+It2 uninitialized_copy_n(shm_ptr_with_raw_ptr_dispatch<T> f, Size n, It2 d)
+{
+  if (n == 0)
+    return d;
+  throw std::runtime_error("shm_ptr uninitialized_copy_n not implemented");
+  // d.wSP_->fence();
```
Comment (markdewing) [inline src/AFQMC/Memory/SharedMemory/shm_ptr_with_raw_ptr_dispatch.hpp:489]: Is the commented-out code a possible implementation?  It looks copied from the previous or following function.

**Comment by ye-luo** [comment]
Test this please

**Review by ye-luo — APPROVED** [review]

**Comment by correaa** [comment]
what is this failure in the CI? 

https://github.com/QMCPACK/qmcpack/actions/runs/6856906594/job/18644992220

**Comment by ye-luo** [comment]
Seems unrelated to this PR. Figuring it out.

**Comment by prckent** [comment]
Unfortunately something has broken in the CI setup for those runs. The machine is fine.

File: src/AFQMC/Memory/SharedMemory/shm_ptr_with_raw_ptr_dispatch.hpp [hunk 2]
```diff
@@ -466,14 +466,14 @@ shm_ptr_with_raw_ptr_dispatch<T> copy(It1 first, It1 last, shm_ptr_with_raw_ptr_
...
-template<class It1, class Size, typename T>
+template<class It1, typename Size, typename T>
 shm_ptr_with_raw_ptr_dispatch<T> uninitialized_copy_n(It1 f, Size n, shm_ptr_with_raw_ptr_dispatch<T> d)
 {
   if (n == 0)
     return d;
   d.wSP_->fence();
   using std::uninitialized_copy_n;
-  if (d.wSP_->get_group().root())
+  if (d.wSP_->get_group().root())
```
Comment (ye-luo) [inline src/AFQMC/Memory/SharedMemory/shm_ptr_with_raw_ptr_dispatch.hpp:476]: trailing white space.

**Review by ye-luo — APPROVED** [review]

**Comment by prckent** [comment]
Test this please

**Pull request merged**